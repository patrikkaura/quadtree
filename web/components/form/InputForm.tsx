import { Constants } from "@core/config";
import LoadingButton from "@mui/lab/LoadingButton";
import Button from "@mui/material/Button";
import Stack from "@mui/material/Stack";
import TextField from "@mui/material/TextField";
import Typography from "@mui/material/Typography";
import { ChangeEvent, useCallback, useMemo } from "react";

export type Props = {
  isDataLoaded: boolean;
  isLoading: boolean;
  sequence: string;
  threshold: number;
  onSequenceChange: (sequence: string) => void;
  onThresholdChange: (threshold: number) => void;
  onDataReset: () => void;
  onAnalyseClick: () => void;
};

export function InputForm({
  isDataLoaded = false,
  isLoading = false,
  sequence,
  threshold,
  onSequenceChange,
  onThresholdChange,
  onDataReset,
  onAnalyseClick,
}: Readonly<Props>) {
  const handleSetSequence = useCallback(
    ({ target: { value } }: ChangeEvent<HTMLInputElement>) => {
      onSequenceChange(value);
    },
    [onSequenceChange]
  );

  const handleSetThreshold = useCallback(
    ({ target: { value } }: ChangeEvent<HTMLInputElement>) => {
      onThresholdChange(parseFloat(value));
    },
    [onThresholdChange]
  );

  const handleResetForm = useCallback(() => {
    onDataReset();
  }, [onDataReset]);

  const inputError = useMemo(() => {
    const sequenceError =
      sequence.length <= Constants.MINIMUM_SEQUENCE_LENGTH ||
      sequence.length >= Constants.MAXIMUM_SEQUENCE_LENGTH;
    const thresholdError =
      threshold <= Constants.MINIMUM_THRESHOLD ||
      threshold >= Constants.MAXIMUM_THRESHOLD;

    return {
      sequence: sequenceError,
      threshold: thresholdError,
    };
  }, [sequence, threshold]);

  return (
    <>
      <Typography variant="h5">Input params</Typography>
      <Stack direction="row" spacing={4} style={{ paddingTop: 30 }}>
        <TextField
          required
          data-testid="sequence-input"
          label="Sequence"
          multiline
          rows={10}
          fullWidth
          value={sequence}
          disabled={isLoading || isDataLoaded}
          onChange={handleSetSequence}
          InputLabelProps={{ shrink: true }}
          helperText={`Sequence length (bp) has to be at least ${Constants.MINIMUM_SEQUENCE_LENGTH}bp. But maximum up to ${Constants.MAXIMUM_SEQUENCE_LENGTH}bp. Current length: ${sequence.length}`}
          error={inputError.sequence}
        />

        <Stack direction="column" spacing={2}>
          <TextField
            required
            data-testid="threshold-input"
            label="Threshold"
            type="number"
            value={threshold}
            disabled={isLoading || isDataLoaded}
            onChange={handleSetThreshold}
            InputLabelProps={{ shrink: true }}
            helperText={
              "Threshold has to be between 0 and 1. The recommended value is 0.2"
            }
            error={inputError.threshold}
          />
          <LoadingButton
            data-testid="start-button"
            color="inherit"
            variant="outlined"
            loading={isLoading}
            onClick={onAnalyseClick}
            disabled={
              inputError.sequence || inputError.threshold || isDataLoaded
            }
          >
            START ANALYSIS
          </LoadingButton>
          <Button
            data-testid="reset-button"
            color="error"
            variant="outlined"
            disabled={isLoading || !isDataLoaded}
            onClick={handleResetForm}
          >
            RESET ANALYSIS
          </Button>
        </Stack>
      </Stack>
    </>
  );
}
