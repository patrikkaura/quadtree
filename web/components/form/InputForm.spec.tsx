import { fireEvent, render, screen } from "@testing-library/react";

import { InputForm, Props } from "./InputForm";

describe("InputForm", () => {
  let props: Props;

  beforeEach(() => {
    props = {
      isDataLoaded: false,
      isLoading: false,
      sequence: "",
      threshold: 0,
      onSequenceChange: jest.fn(),
      onThresholdChange: jest.fn(),
      onDataReset: jest.fn(),
      onAnalyseClick: jest.fn(),
    };
  });

  const renderComponent = ({
    componentProps = props,
  }: {
    componentProps?: Props;
  }) => render(<InputForm {...componentProps} />);

  it("should render", () => {
    const { container } = renderComponent({});
    expect(container).toMatchSnapshot();
  });

  it("should call onSequenceChange when sequence input changes", () => {
    renderComponent({});

    const sequenceInput = screen.getByTestId(
      "sequence-input"
    ) as HTMLTextAreaElement;

    fireEvent.change(sequenceInput, { target: { value: "ATG" } });

    expect(props.onSequenceChange).toHaveBeenCalledWith("ATG");
    expect(props.onSequenceChange).toHaveBeenCalledTimes(1);
  });

  it("should call onThresholdChange when threshold input changes", () => {
    renderComponent({});

    const thresholdInput = screen.getByTestId("threshold-input");

    fireEvent.change(thresholdInput, { target: { value: "0.2" } });

    expect(props.onThresholdChange).toHaveBeenCalledWith(0.2);
    expect(props.onThresholdChange).toHaveBeenCalledTimes(1);
  });

  it("should call onDataReset when reset button clicked", () => {
    renderComponent({
      componentProps: {
        ...props,
        isDataLoaded: true,
      },
    });

    const resetButton = screen.getByTestId("reset-button");

    fireEvent.click(resetButton);

    expect(props.onDataReset).toHaveBeenCalledTimes(1);
    expect(props.onDataReset).toHaveBeenCalledWith();
  });

  it("should show input errors when threshold and sequence hits limits", () => {
    renderComponent({
      componentProps: { ...props, sequence: "ATATA", threshold: -1 },
    });

    expect(
      screen
        .getByTestId("threshold-input")
        // eslint-disable-next-line testing-library/no-node-access
        .querySelector("input")
    ).toMatchSnapshot();
    expect(
      screen
        .getByTestId("sequence-input")
        // eslint-disable-next-line testing-library/no-node-access
        .querySelector("textarea")
    ).toMatchSnapshot();
  });

  it("should disable inputs when isDataLoaded", () => {
    renderComponent({
      componentProps: { ...props, isDataLoaded: true },
    });

    expect(screen.getByTestId("sequence-input")).toBeDisabled();
    expect(screen.getByTestId("threshold-input")).toBeDisabled();
  });

  it("should disable inputs when isLoading", () => {
    renderComponent({
      componentProps: { ...props, isLoading: true },
    });

    expect(screen.getByTestId("sequence-input")).toBeDisabled();
    expect(screen.getByTestId("threshold-input")).toBeDisabled();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });
});
