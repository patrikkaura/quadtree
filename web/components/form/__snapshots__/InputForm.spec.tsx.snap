// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputForm should render 1`] = `
<div>
  <h5
    class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
  >
    Input params
  </h5>
  <div
    class="css-19qjufr-MuiStack-root"
    style="padding-top: 30px;"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      data-testid="sequence-input"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-error Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for=":r0:"
        id=":r0:-label"
      >
        Sequence
        <span
          aria-hidden="true"
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk Mui-error css-wgai2y-MuiFormLabel-asterisk"
        >
           
          *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-error MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-8ewcdo-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <textarea
          aria-describedby=":r0:-helper-text"
          aria-invalid="true"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r0:"
          required=""
          rows="10"
          style="height: 0px; overflow: hidden;"
        />
        <textarea
          aria-hidden="true"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
          readonly=""
          style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
          tabindex="-1"
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-14lo706"
          >
            <span>
              Sequence
               
              *
            </span>
          </legend>
        </fieldset>
      </div>
      <p
        class="MuiFormHelperText-root Mui-error MuiFormHelperText-sizeMedium MuiFormHelperText-contained Mui-required css-1wc848c-MuiFormHelperText-root"
        id=":r0:-helper-text"
      >
        Sequence length (bp) has to be at least 50bp. But maximum up to 10000bp. Current length: 0
      </p>
    </div>
    <div
      class="css-1p5q5e5-MuiStack-root"
    >
      <div
        class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        data-testid="threshold-input"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-error MuiFormLabel-filled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for=":r1:"
          id=":r1:-label"
        >
          Threshold
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk Mui-error css-wgai2y-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-error MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-describedby=":r1:-helper-text"
            aria-invalid="true"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r1:"
            required=""
            type="number"
            value="0"
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-14lo706"
            >
              <span>
                Threshold
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
        <p
          class="MuiFormHelperText-root Mui-error MuiFormHelperText-sizeMedium MuiFormHelperText-contained MuiFormHelperText-filled Mui-required css-1wc848c-MuiFormHelperText-root"
          id=":r1:-helper-text"
        >
          Threshold has to be between 0 and 1. The recommended value is 0.2
        </p>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiLoadingButton-root MuiButton-outlined MuiButton-outlinedInherit MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorInherit Mui-disabled MuiButton-root MuiLoadingButton-root MuiButton-outlined MuiButton-outlinedInherit MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorInherit css-kpoh2l-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
        data-testid="start-button"
        disabled=""
        id=":r2:"
        tabindex="-1"
        type="button"
      >
        START ANALYSIS
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedError MuiButton-sizeMedium MuiButton-outlinedSizeMedium Mui-disabled MuiButton-root MuiButton-outlined MuiButton-outlinedError MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-180lf2n-MuiButtonBase-root-MuiButton-root"
        data-testid="reset-button"
        disabled=""
        tabindex="-1"
        type="button"
      >
        RESET ANALYSIS
      </button>
    </div>
  </div>
</div>
`;

exports[`InputForm should show input errors when threshold and sequence hits limits 1`] = `
<input
  aria-describedby=":rd:-helper-text"
  aria-invalid="true"
  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
  id=":rd:"
  required=""
  type="number"
  value="-1"
/>
`;

exports[`InputForm should show input errors when threshold and sequence hits limits 2`] = `
<textarea
  aria-describedby=":rc:-helper-text"
  aria-invalid="true"
  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
  id=":rc:"
  required=""
  rows="10"
  style="height: 0px; overflow: hidden;"
>
  ATATA
</textarea>
`;
